torch_xla/csrc/xla_lower_util.cpp

```diff
@@ -1194,6 +1194,27 @@ xla::XlaOp BuildCdistForward(xla::XlaOp x1, xla::XlaOp x2, xla::XlaOp p,
   }
 }
 
+xla::XlaOp BuildPixelShuffle(xla::XlaOp input, int64_t upscale_factor) {
+  const xla::Shape& input_shape = ShapeHelper::ShapeOfXlaOp(input);
+  absl::Span<const int64_t> dimensions = input_shape.dimensions();
+  int64_t batch_size = dimensions[0];
+  int64_t channels = dimensions[1];
+  int64_t height = dimensions[2];
+  int64_t width = dimensions[3];
+
+  int64_t new_channels = channels / (upscale_factor * upscale_factor);
+  int64_t new_height = height * upscale_factor;
+  int64_t new_width = width * upscale_factor;
+
+  xla::XlaOp tmp =
+      xla::Reshape(input, {batch_size, new_channels, upscale_factor,
+                           upscale_factor, height, width});
+  tmp = xla::Transpose(tmp, {0, 1, 4, 2, 5, 3});
```

Comment:
Maybe I just don't understand `pixel_shuffle` enough, but why is this `tranpose` needed here?

Reply:
I am using https://g.co/gemini/share/8ee217185f5f. We permute the matrix so that we can easily do reshape next by stacking those dimensions. Same logic was used here: https://github.com/pytorch/xla/blob/786484d2fd271e519c4f222c1224f836322662a8/experimental/torch_xla2/torch_xla2/_ops.py#L987.

Reply:
Nice! LGTM.

Reply:
We permute the matrix so that we can easily do reshape next by stacking those dimensions. Same logic was used here: https://github.com/pytorch/xla/blob/786484d2fd271e519c4f222c1224f836322662a8/experimental/torch_xla2/torch_xla2/_ops.py#L987.